### PR TITLE
[FIX] Refactoring of MRMDecoy

### PR DESCRIPTION
**Pull request by grosenberger**
This pull request addresses the following issues:

* The problem raised in #1769 is now also fixed in OpenSwathDecoyGenerator
* Theoretical transitions are now generated using MRMIonSeries
* Speed-up on a large assay library is  ~3x

### CODE REVIEW

**Comment by hroest** [comment]
@grosenberger is pyOpenMS adopted to the changes?

File: src/tests/class_tests/openms/data/MRMDecoyGenerator_output.TraML [hunk 1]
```diff
@@ -10,16 +10,6 @@
     </Protein>
   </ProteinList>
   <CompoundList>
-    <Peptide id="DECOY_AAAAAK" sequence="AAAAAK">
```
Comment (hroest) [inline src/tests/class_tests/openms/data/MRMDecoyGenerator_output.TraML]: I guess this was the issue: a peptide without transitions?

File: src/tests/class_tests/openms/data/MRMDecoyGenerator_output.TraML [hunk 2]
```diff
@@ -465,12 +455,51 @@
       <cvParam cvRef="MS" accession="MS:1001226" name="product ion intensity" value="55.7"/>
       <cvParam cvRef="MS" accession="MS:1002008" name="decoy SRM transition"/>
     </Transition>
+    <Transition id="DECOY_AQRT.YILAGVENSK.2_y8_1.light" peptideRef="DECOY_YILAGVENSK">
```
Comment (hroest) [inline src/tests/class_tests/openms/data/MRMDecoyGenerator_output.TraML]: why does the PR add new transitions?

File: src/tests/class_tests/openms/source/MRMDecoy_test.cpp [hunk 3]
```diff
@@ -409,168 +396,17 @@ START_SECTION((void generateDecoys(OpenMS::TargetedExperiment & exp,
   traml.load(OPENMS_GET_TEST_DATA_PATH(in), targeted_exp);
 
   MRMDecoy decoys = MRMDecoy();
-  decoys.restrictTransitions(targeted_exp, min_transitions, max_transitions);
```
Comment (grosenberger) [inline src/tests/class_tests/openms/source/MRMDecoy_test.cpp]: restrictTransitions was removed from MRMDecoy. The implementation in MRMAssay should be used if needed. Decoys should always have the exact same transition numbers as targets, or be excluded if this is not possible.

File: src/tests/class_tests/openms/data/MRMDecoyGenerator_output.TraML (hunk 2 above)
Comment (grosenberger) [inline src/tests/class_tests/openms/data/MRMDecoyGenerator_output.TraML]: restrictTransitions() was applied previously in the unit test but is now obsolete. The additional three transitions come from a target assay originally containing 9 transitions, which were filtered to three in the original implementation.

File: src/tests/class_tests/openms/data/MRMDecoyGenerator_output.TraML (hunk 1 above)
Comment (grosenberger) [inline src/tests/class_tests/openms/data/MRMDecoyGenerator_output.TraML]: Exactly. Empty peptides and proteins are now correctly excluded.

**Comment by grosenberger** [comment]
Thanks for the review, @hroest. I added patches for pyOpenMS and did a rebase to the current version.

**Comment by hroest** [comment]
great, I restarted travis but to me it looks like we can merge. great that you now also test the removal of these peptides (even if it was by accident :-)

**Pull request merged**